bpf: Forbid implicit int conversions

pchaigno opened the pull request:
This pull request enables flag `-Wimplicit-int-conversion` in Clang to forbid implicit integer conversions. Enforcing this restriction at the compiler would have allowed us to avoid 1cf3ef3 ("bpf: Fix invalid trace reason in bpf_host") if done earlier. It also allowed us to detect and fix another minor incorrect cast, fixed at https://github.com/cilium/cilium/pull/18429.

To help avoid implicit integer conversions, several lists of constants are converted to enumerations. Flag `-Wenum-conversion` is therefore also enabled. Using enumerations whenever possible also uncovered a bug, fixed in the first commit.

See commits for details.

pchaigno committed:
bpf: Define enum trace_point

Define a trace_point enumeration to hold the TRACE_{FROM,TO}_XXX
constants. Doing so will allow us to rely on the compiler to expose
errors in subsequent commits.

Because of this new enumeration, we also need to define all cases
explicitly in the update_trace_metrics switch. Otherwise, the Clang
errors with the Wswitch warning.

This commit should include no functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Define enum nat_dir

Define a nat_dir enumeration to hold the NAT_DIR_{IN,E}GRESS constants.
Doing so will allow to rely on the compiler to expose errors in subsequent
commits.

This commit should include no functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Define enum ct_dir and metric_dir

Define ct_dir and metric_dir enumerations to hold the
{CT,METRIC}_{INGRESS,EGRESS,SERVICE} constants. Doing so will allow to
rely on the compiler to expose errors, notably if one tries to use a
ct_dir enum as a metric_dir enum without conversion.

This commit should include no functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Define enum ct_status for CT states

This commit defines enumeration ct_status for constants CT_NEW & co.
ct_state was already taken for the structure that holds information from
the conntrack maps. Having a dedicated enumeration helps clarify the
intent of the various variables that hold CT states. It will also allow
us to use the compiler in subsequent commits to detect errors.

This commit shouldn't contain any functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Make LXC_ID 16-bit long

Until now, the LXC_ID static data (i.e., patched in the object file
before loading) was declared and used as a 32-bit integer. The Cilium
endpoint IDs are however encoded on 16 bits and several functions in the
datapath expect a short instead of an integer.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Encode ports on 16 bits in DSR IPv6 options

We use the IPv6 option to pass information for DSR between nodes. The L4
port in there is encoded on 32 bits, which doesn't match the expectation
of the rest of the datapath (16 bits as usual) and leads to implicit int
casts. This commit changes the structure to encode the port on 16 bits.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Explicit proxy port variable

The policy layer (policy_can_access_ingress & co) returns an integer
which can take different values:
- CTX_ACT_OK the packet is allowed based only on L3 and L4.
- A negative error code if the packet is rejected.
- A positive integer if the packet should be redirected to the proxy.

In that last case, the positive integer represents a proxy port, which
is encoded on only 16 bits. This commit introduces a 16-bits variable
to hold the proxy port, to explicit the int conversion.

This commit should include no functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf/Makefile: Forbid implicit int conversions

This commit forbids implicit int conversions as well as conversions
between enumerations. A bunch of int conversions are made explicit as
well, but there shouldn't be any functional changes.

Enforcing this restriction at the compiler would have allowed us to
avoid 1cf3ef3f ("bpf: Fix invalid trace reason in bpf_host") if done
earlier. It allowed us to detect and fix another minor incorrect cast,
sent before this patch and merged as 108aa421 ("bpf: Fix implicit cast
for BPF TPROXY debug message").

Several implicit int conversions "fixed" here are inherited from u32 BPF
APIs (holding <32bit values for historic reasons). Other are due to our
use of 32-bit long slots for the metadata. Finally, others are due to
the get_prandom_u32 BPF helper returning a 32-bit integer, which can of
course be safely used as a smaller integer.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
elf: Rename symbolUint32 to symbolData

With the next commit, this symbol type will include 16-bit long symbols
as well so we need to adapt the name.

Signed-off-by: Paul Chaignon <paul@cilium.io>

nathanjsweet commented on bpf/lib/trace.h:
```diff
@@ -110,6 +103,19 @@ update_trace_metrics(struct __ctx_buff *ctx, __u8 obs_point, enum trace_reason r
+	 * we have separate L7 proxy "forwarded" and "dropped" (ingress/egress)
+	 * counters in the proxy layer to capture these metrics.
+	 */
+	case TRACE_FROM_LXC:
```
Why add the cases explicitly?

pchaigno replied:
Good question. That's one of the upside of using an enumeration for `obs_point`. As soon as we do that, LLVM is able to detect that several switch cases are missing:
```
error: enumeration values 'TRACE_TO_PROXY', 'TRACE_FROM_LXC', and 'TRACE_FROM_PROXY' not handled in switch [-Werror,-Wswitch]
        switch (obs_point) {
                ^
1 error generated.
```
This warning (`-Wswitch`) is enabled by default and turned into an error because we use `-Werror`.

I'll add a note in the commit description. I need to rebase anyway.

pchaigno replied:
I've moved this code change to the next commit, as it seemed more appropriate. Thanks for raising this Nathan! :pray:

nathanjsweet reviewed (approved):
Very cool.

pchaigno committed:
bpf: Fix missing metric updates on from-{stack,host}

Function update_trace_metrics updates metrics (number of bytes and
packets) at various observation points. Observation points
TRACE_FROM_HOST and TRACE_FROM_STACK were however missing from the list.

This oversight was identified thanks to the new enumeration trace_point
introduced in the next commit.

Signed-off-by: Paul Chaignon <paul@cilium.io>
